### PR TITLE
smc-abc backward_val

### LINKED ISSUE #3714: Bounded prior causes Value Error in sample_smc(kernel='ABC')
Doing ```sample_smc(kernel='ABC')``` and setting one particular prior from ```Normal``` to the bounded ```Uniform``` causes an error ```setting an array element with a sequence.```

**Please provide a minimal, self-contained, and reproducible example.**
The gist is here: https://gist.github.com/aakhmetz/372d0bb18ea89882437d6c63053f45b8
```python
def _sim(gtmean, gtsd, σ1, ρ, w):
    weights = np.random.dirichlet(w, 1)[0]
    draws = np.random.multinomial(len(ds), weights, size=1)[0]
    # generation time
    α = (gtmean/gtsd)**2
    β = gtsd**2/gtmean
    gt = ss.gamma(α, scale = β)
    # incubation period
    incmean = 12.3; incvar = 12.1
    incα = (incmean**2)/incvar
    incβ = incvar/incmean
    ht = ss.gamma(incα, scale = incβ)
    # Mechanism 1
    res = np.random.normal(0, σ1, draws[0])
    # Mechanism 2
    μ2 = abs(np.random.gamma(α, β, draws[1]) - np.random.gamma(α, β, draws[1]))
    res = np.r_[res,np.random.normal(μ2, 0.5, size=draws[1])]
    # Mechanism 3
    ## incorporating copulas
    ### see https://twiecki.io/blog/2018/05/03/copulas/ for details
    if (draws[2]>=1):
        x = np.random.multivariate_normal([0, 0], [[1., ρ], [ρ, 1.]], draws[2])
        norm = ss.distributions.norm()
        x_unif = norm.cdf(x)
        τ = ht.ppf(x_unif[:, 0]) 
        u = gt.ppf(x_unif[:, 1])
        μ3 = abs(gt.rvs(draws[2]) + u - τ)
        res = np.r_[res, np.random.normal(μ3, 0.5, size=draws[2])]
    # output
    return np.sort(res)

with pm.Model():
    # Generation time interval
    gtmean = pm.HalfNormal('gtmean', sd=20) 
    gtsd = pm.HalfNormal('gtsd', sd=5)
    # for Mechanism 1
    σ1 = pm.HalfNormal('σ1', sd=5, testval=5)
    # corr.coeff. b/w incubation period and serial interval
    ρ = pm.Uniform('ρ', 0, 1, testval=0) ## <- does not work
    ## ρ = pm.Bound(pm.Normal,lower=-1,upper=1)('ρ', 0, 1, testval=0) ## <- does not work
    ## ρ = pm.Normal('ρ', 0, .1, testval=-.5) <- previous variant
    
    w = pm.Dirichlet('w', a=np.array([1, 5, 1]))
    s = pm.Simulator('s', _sim, observed=np.sort([x+0.5 for x in ds]))
    trace = pm.sample_smc(1000, kernel="ABC", epsilon=0.1, parallel=True)
```

**Please provide the full traceback.**
```python
Sample initial stage: ...
---------------------------------------------------------------------------
RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/aakhmetz/anaconda3/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/aakhmetz/anaconda3/lib/python3.6/multiprocessing/pool.py", line 47, in starmapstar
    return list(itertools.starmap(args[0], args[1]))
  File "/home/aakhmetz/anaconda3/lib/python3.6/site-packages/pymc3/smc/smc.py", line 438, in __call__
    sim_data = self.function(**func_parameters)
  File "<ipython-input-16-1302d6d723b0>", line 22, in _sim
    x = np.random.multivariate_normal([0, 0], [[1., ρ], [ρ, 1.]], draws[2])
  File "mtrand.pyx", line 3758, in numpy.random.mtrand.RandomState.multivariate_normal
ValueError: setting an array element with a sequence.
"""

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<timed exec> in <module>

~/anaconda3/lib/python3.6/site-packages/pymc3/smc/sample_smc.py in sample_smc(draws, kernel, n_steps, parallel, start, cores, tune_steps, p_acc_rate, threshold, epsilon, dist_func, sum_stat, progressbar, model, random_seed)
    136     smc.initialize_population()
    137     smc.setup_kernel()
--> 138     smc.initialize_logp()
    139 
    140     while smc.beta < 1:

~/anaconda3/lib/python3.6/site-packages/pymc3/smc/smc.py in initialize_logp(self)
    142             priors = self.pool.starmap(self.prior_logp, [(sample,) for sample in self.posterior])
    143             likelihoods = self.pool.starmap(
--> 144                 self.likelihood_logp, [(sample,) for sample in self.posterior]
    145             )
    146         else:

~/anaconda3/lib/python3.6/multiprocessing/pool.py in starmap(self, func, iterable, chunksize)
    272         `func` and (a, b) becomes func(a, b).
    273         '''
--> 274         return self._map_async(func, iterable, starmapstar, chunksize).get()
    275 
    276     def starmap_async(self, func, iterable, chunksize=None, callback=None,

~/anaconda3/lib/python3.6/multiprocessing/pool.py in get(self, timeout)
    642             return self._value
    643         else:
--> 644             raise self._value
    645 
    646     def _set(self, i, obj):

ValueError: setting an array element with a sequence.
```

**Please provide any additional information below.**
Because I am new in SMC-ABC, I maybe be wrong in something. Although it seems to me that my code is legit.

## Versions and main components

* PyMC3 Version: 3.8
* Theano Version: 1.0.4
* Python Version: 3.6.9
* Operating system: ubuntu
* How did you install PyMC3: (conda/pip) latest master version from github

### PR DESCRIPTION
**Pull request by aloctavodia**
fix #3714 and add a few small tweaks while been there :-)

### CODE REVIEW

File: pymc3/distributions/transforms.py [hunk 1]
```diff
@@ -265,8 +263,8 @@ class Interval(ElemwiseTransform):
     def __init__(self, a, b):
         self.a = tt.as_tensor_variable(a)
         self.b = tt.as_tensor_variable(b)
-        self.a_ = a
-        self.b_ = b
+        self.a_ = self.a.eval()
```
Comment (junpenglao) [inline pymc3/distributions/transforms.py]: This would break if `a` is a random variable  with upstream dependence right?

Comment (aloctavodia) [reply pymc3/distributions/transforms.py]: maybe that's why I am seeing a lot of errors in the tests :-) I will check it ASAP

Comment (junpenglao) [reply pymc3/distributions/transforms.py]: Yeah generally `{tensor}.eval()` is not very reliable as most of the case they have some dependencies.

Comment (aloctavodia) [reply pymc3/distributions/transforms.py]: I tried a couple of alternatives, this is the best one. Still a couple of errors. Any suggestion?

Comment (junpenglao) [reply pymc3/distributions/transforms.py]: draw_value would be how I approach it as well, but if `a` is a random variable you will still have problem with `point=None`. Maybe @lucianopaz has some ideas.

**Comment by aloctavodia** [comment]
Close in favour of #3744

**Pull request closed without merging**